File: javascript/smartlist/smartlist-api-mysql-project/src/modules/candidate/candidate-question-strategy/question-strategy-factory.ts
import { Question } from 'src/entities/question.entity';
import { QuestionType } from 'src/enums/question.type';
import { SubmitAnswerReqDto } from '../dto/request/submit-answer.req.dto';
import { McqMultipleStrategy } from './mcq-multiple-strategy';
import { McqSingleWeightedSelectStrategy } from './mcq-single-weighted-select-strategy';
import { McqSingleStrategy } from './mcq-single-strategy';
import { SurveyStrategy } from './survey-strategy';
import { VideoStrategy } from './video-strategy';
import { TextStrategy } from './text-strategy';
import { IQuestionStrategy } from '../interfaces/interface-question-strategy';
import { FileStrategy } from './file-strategy';

export function getQuestionStrategy(
  submitAnswerReqDto: SubmitAnswerReqDto,
  question: Question,
): IQuestionStrategy {
  let questionStrategy: IQuestionStrategy;
  switch (question.type) {
    case QuestionType.SCORED_MCQ_SINGLE:
      questionStrategy = new McqSingleStrategy(submitAnswerReqDto, question);
      break;
    case QuestionType.SCORED_MCQ_SINGLE_WEIGHTED_SELECT:
      questionStrategy = new McqSingleWeightedSelectStrategy(
        submitAnswerReqDto,
        question,
      );
      break;
    case QuestionType.SCORED_MCQ_MULTIPLE:
      questionStrategy = new McqMultipleStrategy(submitAnswerReqDto, question);
      break;
    case QuestionType.SURVEY:
      questionStrategy = new SurveyStrategy(submitAnswerReqDto, question);
      break;
    case QuestionType.TEXT_RESPONSE:
      questionStrategy = new TextStrategy(submitAnswerReqDto, question);
      break;
    case QuestionType.VIDEO_RESPONSE:
      questionStrategy = new VideoStrategy(submitAnswerReqDto, question);
      break;
    case QuestionType.FILE_RESPONSE:
      questionStrategy = new FileStrategy(submitAnswerReqDto, question);
      break;
    default:
      break;
  }
  return questionStrategy;
}

File: java/smartlist/src/main/java/com/smartlist/module/candidate/questionstrategy/QuestionStrategyFactory.java
package com.smartlist.module.candidate.questionstrategy;

import com.smartlist.model.Question;
import com.smartlist.module.candidate.dto.request.SubmitAnswerReqDTO;
import com.smartlist.module.candidate.interfaces.IQuestionStrategy;
import com.smartlist.module.file.FileRepository;
import com.smartlist.module.question.AnswerOptionRepository;
import org.springframework.stereotype.Service;

@Service
public class QuestionStrategyFactory {

    private final AnswerOptionRepository answerOptionRepository;
    private final FileRepository fileRepository;

    public QuestionStrategyFactory(final AnswerOptionRepository answerOptionRepository, final FileRepository fileRepository){
        this.answerOptionRepository = answerOptionRepository;
        this.fileRepository = fileRepository;
    }

    public IQuestionStrategy getQuestionStrategy(final SubmitAnswerReqDTO submitAnswerReqDto, final Question question){

        IQuestionStrategy questionStrategy = null;

        switch (question.getType()) {
            case SCORED_MCQ_SINGLE -> questionStrategy = new McqSingleStrategy(submitAnswerReqDto, question, this.answerOptionRepository);
            case SCORED_MCQ_SINGLE_WEIGHTED_SELECT -> questionStrategy = new McqSingleWeightedStrategy(submitAnswerReqDto, question, this.answerOptionRepository);
            case SCORED_MCQ_MULTIPLE -> questionStrategy = new McqMultipleStrategy(submitAnswerReqDto, question, this.answerOptionRepository);
            case SURVEY -> questionStrategy = new SurveyStrategy(submitAnswerReqDto, question, this.answerOptionRepository);
            case TEXT_RESPONSE -> questionStrategy = new TextStrategy(submitAnswerReqDto, question, this.answerOptionRepository);
            case VIDEO_RESPONSE -> questionStrategy = new VideoStrategy(submitAnswerReqDto, question, this.answerOptionRepository, this.fileRepository);
            case FILE_RESPONSE -> questionStrategy = new FileStrategy(submitAnswerReqDto, question, this.answerOptionRepository, this.fileRepository);
        }
        return questionStrategy;

    }
}
